File: 678-valid-parenthesis-string/valid-parenthesis-string.py
class Solution:
    def checkValidString(self, s: str) -> bool:

        openstack = []
        asterisk = []

        for i, ch in enumerate(s):
            if ch=='(':
                openstack.append(i)
            elif ch=='*':
                asterisk.append(i)
            else:

                if openstack:
                    openstack.pop()
                elif asterisk:
                    asterisk.pop()
                    # asterisk idx that came before this ')'
                    # asterisk can substiture open bracket
                else:
                    return False
    
        while openstack and asterisk:
            if openstack[-1] > asterisk[-1]:
                # if open bracket came after close bracket
                #  we cant substitute asterisk as "close bracket"
                 return False
            else:
                openstack.pop() 
                asterisk.pop()
        
        return len(openstack)==0

File: 678-valid-parenthesis-string/valid-parenthesis-string.java
class Solution {
    public boolean checkValidString(String s) {

        Stack<Integer> open_index = new Stack<>();
        Stack<Integer> star_idx = new Stack<>();
        

        for(int i = 0; i < s.length(); i++) {
            Character ch = s.charAt(i);
            if(ch == '*') {
                star_idx.add(i);
            }
            else if (ch == '(') {
                open_index.add(i);
            }
            else  {
                if (!open_index.isEmpty()) {
                    open_index.pop();
                }
                else if (!star_idx.isEmpty()){
                    star_idx.pop();
                }
                else {
                    return false;
                }
            }     
        }

        // System.out.print(open_index);
        // System.out.println(star_idx);
        
        while(!open_index.isEmpty() && !star_idx.isEmpty()) {
            if(open_index.pop() > star_idx.pop()) {
                // System.out.print(open_index);
                // System.out.println(star_idx);
                return false;  // Star comes before '(', can't use as ')'
            } 
        }
    
    return open_index.isEmpty();
    }
} 